File: Data Structures & Algorithms/combination-target-sum-ii/submission-0.cpp
class Solution {
public:
    vector<vector<int>> res;
    unordered_map<int, int> count;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> curr;
        vector<int> nums;
        for (int candidate : candidates) {
            if (!count[candidate]) {
                nums.push_back(candidate);
            }
            count[candidate] += 1;
        }
        search(nums, target, curr, 0);
        return res;
    }
    void search(vector<int>& nums, int target, vector<int>& curr, int i) {
        if (target == 0) {
            res.push_back(curr);
            return;
        }
        if (target < 0 || i >= nums.size()) {
            return;
        }

        if (count[nums[i]]) {
            curr.push_back(nums[i]);
            count[nums[i]]--;
            search(nums, target - nums[i], curr, i);
            curr.pop_back();
            count[nums[i]]++;
        }

        search(nums, target, curr, i + 1);
    }
};

File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    def binary_search(self, left, right, nums, target):
        if left > right:
            return -1
        
        m = (left + right) // 2

        if nums[m] == target:
            return m 
        elif nums[m] < target:
            return self.binary_search(m+1, right, nums, target)
        else:
            return self.binary_search(left, m-1, nums, target)


    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        return self.binary_search(left, right, nums, target)

        


            
        
            
        